<a href="https://colab.research.google.com/github/sagar2582/World-Bank-Binary-classifier/blob/main/World_Bank_Binary_Classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
import matplotlib.pyplot as plt 
import numpy as np 
import os # accessing directory structure
import pandas as pd 

In [6]:
# Connecting to Drive to import the dataset

from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [67]:
df = pd.read_csv('/content/drive/My Drive/Taiyo/world_bank_dataset.csv')

In [68]:
# Creating a copy of our dataset for futher use

df_copy = df.copy()

In [69]:
# First look
df.head()

id                    regionname  \
0  P151169  Middle East and North Africa   
1  P154943         East Asia and Pacific   
2  P161382         East Asia and Pacific   
3  P163922         East Asia and Pacific   
4  P169279         East Asia and Pacific   

                                         countryname prodline  \
0              Kingdom of Morocco;Kingdom of Morocco       PE   
1                  Kingdom of Tonga;Kingdom of Tonga       PE   
2  Republic of the Marshall Islands;Republic of t...       PE   
3  Federated States of Micronesia;Federated State...       PE   
4                                        Samoa;Samoa       PE   

                   lendinginstr lendinginstrtype envassesmentcategorycode  \
0  Investment Project Financing               IN                        B   
1  Investment Project Financing               IN                        C   
2  Investment Project Financing               IN                        B   
3  Investment Project Financing               IN                        B   
4  Investment Project Financing               IN                      NaN   

  supplementprojectflg productlinetype projectstatusdisplay  ... mjtheme3name  \
0                    N               L               Active  ...          NaN   
1                    N               L               Active  ...          NaN   
2                    N               L               Active  ...          NaN   
3                    N               L               Active  ...          NaN   
4                    Y               L               Active  ...          NaN   

  mjtheme4name mjtheme5name location GeoLocID GeoLocName Latitude Longitude  \
0          NaN          NaN      NaN      NaN        NaN      NaN       NaN   
1          NaN          NaN      NaN      NaN        NaN      NaN       NaN   
2          NaN          NaN      NaN      NaN        NaN      NaN       NaN   
3          NaN          NaN      NaN      NaN        NaN      NaN       NaN   
4          NaN          NaN      NaN      NaN        NaN      NaN       NaN   

  Country Unnamed: 56  
0     NaN         NaN  
1     NaN         NaN  
2     NaN         NaN  
3     NaN         NaN  
4     NaN         NaN  

[5 rows x 57 columns]

In [70]:
df.tail()

id                   regionname  \
18836  P006578  Latin America and Caribbean   
18837  P037451      Europe and Central Asia   
18838  P037362      Europe and Central Asia   
18839  P037452      Europe and Central Asia   
18840  P037383      Europe and Central Asia   

                                             countryname prodline  \
18836                Republic of Chile;Republic of Chile       PE   
18837  Grand Duchy of Luxembourg;Grand Duchy of Luxem...       PE   
18838              Kingdom of Denmark;Kingdom of Denmark       PE   
18839  Kingdom of the Netherlands;Kingdom of the Neth...       PE   
18840                    French Republic;French Republic       PE   

                     lendinginstr lendinginstrtype envassesmentcategorycode  \
18836    Specific Investment Loan               IN                      NaN   
18837  Structural Adjustment Loan               AD                      NaN   
18838  Structural Adjustment Loan               AD                      NaN   
18839  Structural Adjustment Loan               AD                      NaN   
18840  Structural Adjustment Loan               AD                      NaN   

      supplementprojectflg productlinetype projectstatusdisplay  ...  \
18836                    N               L               Closed  ...   
18837                    N               L               Closed  ...   
18838                    N               L               Closed  ...   
18839                    N               L               Closed  ...   
18840                    N               L               Closed  ...   

      mjtheme3name mjtheme4name mjtheme5name location GeoLocID GeoLocName  \
18836          NaN          NaN          NaN      NaN      NaN        NaN   
18837          NaN          NaN          NaN      NaN      NaN        NaN   
18838          NaN          NaN          NaN      NaN      NaN        NaN   
18839          NaN          NaN          NaN      NaN      NaN        NaN   
18840          NaN          NaN          NaN      NaN      NaN        NaN   

      Latitude Longitude Country Unnamed: 56  
18836      NaN       NaN     NaN         NaN  
18837      NaN       NaN     NaN         NaN  
18838      NaN       NaN     NaN         NaN  
18839      NaN       NaN     NaN         NaN  
18840      NaN       NaN     NaN         NaN  

[5 rows x 57 columns]

In [71]:
df.shape

(18841, 57)

# Dataset Information
The shape of our dataset is (18841, 57)

id --> unique id-number for every project i.e., 18841 unique id's

regionname --> points towards the particular region of the world where the project belongs from

countryname --> name of the country

prodline --> product line - PE or RE (No idea what's this. I let the algo work on it)

lendinginstr --> different investment purposes

lendinginstrtype --> type of investment

envassesmentcategorycode --> investment category code

supplementprojectflg --> (No idea what's this) binary categoriacal column

productlinetype --> Single variable column. No use

projectstatusdisplay --> current live status

**status** --> overall status (Target Variable)

and more .....

## I see a lot of columns with a lot of NULL values, so let's start with cleaning our data and making it ready for visualization

In [72]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18841 entries, 0 to 18840
Data columns (total 57 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        18841 non-null  object 
 1   regionname                18841 non-null  object 
 2   countryname               18841 non-null  object 
 3   prodline                  18841 non-null  object 
 4   lendinginstr              18580 non-null  object 
 5   lendinginstrtype          18580 non-null  object 
 6   envassesmentcategorycode  12731 non-null  object 
 7   supplementprojectflg      18792 non-null  object 
 8   productlinetype           18841 non-null  object 
 9   projectstatusdisplay      18839 non-null  object 
 10  status                    18839 non-null  object 
 11  project_name              18841 non-null  object 
 12  boardapprovaldate         17172 non-null  object 
 13  board_approval_month      17172 non-null  object 
 14  closin

In [73]:
df.describe(include='all')

id regionname                          countryname prodline  \
count     18841      18841                                18841    18841   
unique    18841          7                                  206       10   
top     P151169     Africa  Republic of India;Republic of India       PE   
freq          1       5900                                  806    15119   
mean        NaN        NaN                                  NaN      NaN   
std         NaN        NaN                                  NaN      NaN   
min         NaN        NaN                                  NaN      NaN   
25%         NaN        NaN                                  NaN      NaN   
50%         NaN        NaN                                  NaN      NaN   
75%         NaN        NaN                                  NaN      NaN   
max         NaN        NaN                                  NaN      NaN   

                    lendinginstr lendinginstrtype envassesmentcategorycode  \
count                      18580            18580                    12731   
unique                        18                4                        6   
top     Specific Investment Loan               IN                        B   
freq                        7702            16399                     5738   
mean                         NaN              NaN                      NaN   
std                          NaN              NaN                      NaN   
min                          NaN              NaN                      NaN   
25%                          NaN              NaN                      NaN   
50%                          NaN              NaN                      NaN   
75%                          NaN              NaN                      NaN   
max                          NaN              NaN                      NaN   

       supplementprojectflg productlinetype projectstatusdisplay  ...  \
count                 18792           18841                18839  ...   
unique                    2               1                    4  ...   
top                       N               L               Closed  ...   
freq                  16831           18841                14625  ...   
mean                    NaN             NaN                  NaN  ...   
std                     NaN             NaN                  NaN  ...   
min                     NaN             NaN                  NaN  ...   
25%                     NaN             NaN                  NaN  ...   
50%                     NaN             NaN                  NaN  ...   
75%                     NaN             NaN                  NaN  ...   
max                     NaN             NaN                  NaN  ...   

       mjtheme3name mjtheme4name mjtheme5name  \
count           0.0          0.0          0.0   
unique          NaN          NaN          NaN   
top             NaN          NaN          NaN   
freq            NaN          NaN          NaN   
mean            NaN          NaN          NaN   
std             NaN          NaN          NaN   
min             NaN          NaN          NaN   
25%             NaN          NaN          NaN   
50%             NaN          NaN          NaN   
75%             NaN          NaN          NaN   
max             NaN          NaN          NaN   

                                                 location  \
count                                                4569   
unique                                               4085   
top     0000281132!$!Gaza Strip!$!31.41667!$!34.333328...   
freq                                                   13   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                       

## Going through the target variable

In [74]:
df_copy['status'].value_counts()

Closed      14625
Active       2644
Dropped       991
Pipeline      579
Name: status, dtype: int64

In [75]:
df_copy.shape

(18841, 57)

### According to our problem statement we need to make a binary classifier that will predict weather the project will be 'Closed' or 'Dropped'. Hence' dropping rows with the cases other than above two. It will also reduce our NULL values to some extent

In [76]:
df_copy = df_copy[(df_copy['status'] == 'Closed') | (df_copy['status'] == 'Dropped')]
df_copy.shape

(15616, 57)

### Handling missing Data

In [77]:
# Missing Value Count Function

def show_missing():
    missing = df.columns[df.isnull().any()].tolist()
    return missing

# Missing data counts and percentage
print('Missing Data Count')
print(df[show_missing()].isnull().sum().sort_values(ascending = False))
print('--'*50)
print('Missing Data Percentage')
print(round(df[show_missing()].isnull().sum().sort_values(ascending = False)/len(df)*100,2))

Missing Data Count
mjsector4                   18841
majorsector_percent         18841
mjtheme5name                18841
mjtheme4name                18841
mjtheme3name                18841
mjtheme2name                18841
mjtheme1name                18841
theme                       18841
mjsector5                   18841
mjsector3                   18841
mjsector2                   18841
mjsector1                   18841
Unnamed: 56                 18841
projectdoc                  18841
sector5                     16071
theme5                      15906
Country                     14390
sector4                     14361
GeoLocName                  14273
Longitude                   14272
Latitude                    14272
GeoLocID                    14272
location                    14272
theme4                      13594
sector3                     12169
financier                   11642
theme3                      11333
sector2                      9031
theme2                       

## Removing Features with over 40% missing values

In [78]:
# Now working on df_copy

df_copy = df_copy.drop(['mjsector4','mjtheme5name','mjtheme4name','mjtheme3name','mjtheme2name','mjtheme1name','mjsector5','mjsector3',
                        'mjsector2','mjsector1','Unnamed: 56','sector5','theme5','Country','sector4','GeoLocName','Longitude','Latitude','GeoLocID',
                        'location','theme4','sector3','financier','theme3','sector2','theme2','projectdoc ','theme ','majorsector_percent '], axis = 1)

In [79]:
# Missing Value Count Function

def show_missing():
    missing = df_copy.columns[df_copy.isnull().any()].tolist()
    return missing

# Missing data counts and percentage
print('Missing Data Count')
print(df_copy[show_missing()].isnull().sum().sort_values(ascending = False))
print('--'*50)
print('Missing Data Percentage')
print(round(df_copy[show_missing()].isnull().sum().sort_values(ascending = False)/len(df_copy)*100,2))

Missing Data Count
envassesmentcategorycode    5484
goal                        5324
impagency                   5116
borrower                    5029
closingdate                 2402
boardapprovaldate           1084
board_approval_month        1084
lendinginstr                 208
lendinginstrtype             208
sector                       141
mjsector                     141
lendprojectcost               57
supplementprojectflg          45
dtype: int64
----------------------------------------------------------------------------------------------------
Missing Data Percentage
envassesmentcategorycode    35.12
goal                        34.09
impagency                   32.76
borrower                    32.20
closingdate                 15.38
boardapprovaldate            6.94
board_approval_month         6.94
lendinginstr                 1.33
lendinginstrtype             1.33
sector                       0.90
mjsector                     0.90
lendprojectcost              0.37
supple

In [80]:
df_copy.columns

Index(['id', 'regionname', 'countryname', 'prodline', 'lendinginstr',
       'lendinginstrtype', 'envassesmentcategorycode', 'supplementprojectflg',
       'productlinetype', 'projectstatusdisplay', 'status', 'project_name',
       'boardapprovaldate', 'board_approval_month', 'closingdate',
       'lendprojectcost', 'ibrdcommamt', 'idacommamt', 'totalamt', 'grantamt',
       'borrower', 'impagency', 'url', 'sector1', 'sector', 'mjsector',
       'theme1', 'goal'],
      dtype='object')

Further Inspecting each column and checking the relevance for our result

## **goal**

In [81]:
df_copy['goal'].unique()

array([nan,
       'Global Public Goods Priorities|Millennium Development Goals;Global Public Goods Priorities|Millennium Development Goals;GPG|MDG;Corporate Advocacy Priorities;CAP;Global Public Goods Priorities|Millennium Development Goals;GPG|MDG',
       'Corporate Advocacy Priorities;Corporate Advocacy Priorities;CAP;Corporate Advocacy Priorities;CAP;Global Public Goods Priorities|Corporate Advocacy Priorities;GPG|CAP;Corporate Advocacy Priorities;CAP;Corporate Advocacy Priorities|Millennium Development Goals;CAP|MDG',
       ...,
       'Corporate Advocacy Priorities;Corporate Advocacy Priorities;CAP;Corporate Advocacy Priorities;CAP;Corporate Advocacy Priorities|Millennium Development Goals;CAP|MDG;Global Public Goods Priorities|Millennium Development Goals;GPG|MDG;Corporate Advocacy Priorities;CAP',
       'Corporate Advocacy Priorities|Millennium Development Goals;Corporate Advocacy Priorities|Millennium Development Goals;CAP|MDG;Corporate Advocacy Priorities;CAP;Corporate Adv

This column does not contain any relevance to out target variable and hence will be removed

In [82]:
df_copy = df_copy.drop(['goal'], axis =1)

## **impagency**

This column contains different agencies codes, and can be important in making decisions, thus imputing this column with 'mean'.

In [83]:
df_copy['impagency'].value_counts()

IIST                          3261
T                              284
EET                            247
EPATET                         144
I I                             95
                              ... 
USHAE                            1
UAI ET                           1
LEAL LE                          1
C HI                             1
RCPG                IPMENT       1
Name: impagency, Length: 3183, dtype: int64

In [84]:
# Imputing with mode, as IIST agency appeared 12 times more than any observation

df_copy['impagency'] = df_copy['impagency'].fillna(df_copy['impagency'].mode()[0])
print('impagency Missing values After Imputing:', df_copy['impagency'].isnull().sum())

impagency Missing values After Imputing: 0


## **borrower**

In [85]:
df_copy['borrower'].value_counts()

EET        3159
EPULIC     1344
IIST        456
I           263
T           255
           ... 
LAE CHA       1
IITA A        1
HUSI A        1
SHA P         1
HEP           1
Name: borrower, Length: 1522, dtype: int64

In [86]:
# Imputing with mode, as EET agency appeared more than any observation

df_copy['borrower'] = df_copy['borrower'].fillna(df_copy['borrower'].mode()[0])
print('borrower Missing values After Imputing:', df_copy['borrower'].isnull().sum())

borrower Missing values After Imputing: 0


## **envassesmentcategorycode**

In [87]:
df_copy['envassesmentcategorycode'].value_counts()

C    4561
B    4028
A     745
U     501
F     272
D      25
Name: envassesmentcategorycode, dtype: int64

In [88]:
# making a seperate category for this

df_copy['envassesmentcategorycode'] = df_copy['envassesmentcategorycode'].replace(np.nan, 'E', regex = True)
print(' Missing values After Imputing:', df_copy['envassesmentcategorycode'].isnull().sum())

 Missing values After Imputing: 0


## **closingdate**


### I am assuming that closing dates have no relation to weather a project will be completed and closed or dropped. 

## Hence; I will be dropping the **closingdate** column 


## **boardapprovaldate** & **board_approval_month** also do not hold meaning to the answer we are looking for. Thus; dropping them too.*italicised text*

In [89]:
df_copy = df_copy.drop(['closingdate','boardapprovaldate','board_approval_month'], axis =1)

## **lendinginstr**  &  **lendinginstrtype** both have 208 missing values

In [90]:
df_copy['lendinginstr'].value_counts()

Specific Investment Loan                   7634
Technical Assistance Loan                  1633
Sector Investment and Maintenance Loan     1269
Investment Project Financing                878
Development Policy Lending                  829
Financial Intermediary Loan                 813
Emergency Recovery Loan                     569
Structural Adjustment Loan                  548
Adaptable Program Loan                      515
Sector Adjustment Loan                      368
Learning and Innovation Loan                177
Programmatic Structural Adjustment Loan      52
Poverty Reduction Support Credit             40
UNIDENTIFIED                                 34
Rehabilitation Loan                          23
Debt and Debt Service Reduction Loan         13
Program-for-Results Financing                11
Special Structural Adjustment Loan            2
Name: lendinginstr, dtype: int64

In [91]:
df_copy['lendinginstrtype'].value_counts()

IN    13488
AD     1875
XX       34
PR       11
Name: lendinginstrtype, dtype: int64

### Imputing both of them with mode

In [92]:
df_copy['lendinginstr'] = df_copy['lendinginstr'].fillna(df_copy['lendinginstr'].mode()[0])
print('lendinginstr Missing values After Imputing:', df_copy['lendinginstr'].isnull().sum())

print('--'*50)

df_copy['lendinginstrtype'] = df_copy['lendinginstrtype'].fillna(df_copy['lendinginstrtype'].mode()[0])
print('lendinginstrtype Missing values After Imputing:', df_copy['lendinginstrtype'].isnull().sum())

lendinginstr Missing values After Imputing: 0
----------------------------------------------------------------------------------------------------
lendinginstrtype Missing values After Imputing: 0


### **sector** and **mjsector** both have 141 missing values. Let's take a look

In [93]:
len(df_copy['sector'].value_counts())

4939

In [94]:
df_copy['mjsector'].value_counts()

Agriculture; Fishing and Forestry;Agriculture; Fishing and Forestry                                                                                                                                 1649
Transportation;Transportation                                                                                                                                                                       1191
Public Administration;Public Administration                                                                                                                                                          733
(Historic)Electric Power & Other Energy;(Historic)Electric Power & Other Energy                                                                                                                      583
Financial Sector;Financial Sector                                                                                                                                                                   

In [101]:
df_copy['sector'] = df_copy['sector'].fillna(df_copy['sector'].mode()[0])
print('sector Missing values After Imputing:', df_copy['sector'].isnull().sum())

print('--'*50)

df_copy['mjsector'] = df_copy['mjsector'].fillna(df_copy['mjsector'].mode()[0])
print('mjsector Missing values After Imputing:', df_copy['mjsector'].isnull().sum())

sector Missing values After Imputing: 0
----------------------------------------------------------------------------------------------------
mjsector Missing values After Imputing: 0


## Dropping rest of the missing values as they are not even 0.37 percent of the data we have

In [111]:
df_copy = df_copy.dropna() 

## Now before getting into feature engineering, let's find out the most important features of all. This time I will be using an ExtraTree Classifier